design_notes/v2.md

```diff
@@ -0,0 +1,176 @@
+# VirtualiZarr V2 architecture discussion
+
+## Goals
+
+- Extensible and easily understandable entry-point system for new backends
+- Modular code base that uses Obspec for the object storage interface abstraction, Zarr as the core data model, and Xarray for the concatenation and merging of Zarr stores
+
+## Related issues
+
+- [Fragility of url auto-parsing logic](https://github.com/zarr-developers/VirtualiZarr/issues/561)
+- [Function definition for open_virtual_dataset in next major release](https://github.com/zarr-developers/VirtualiZarr/issues/553)
+- [Support passing configuration options to default_object_store](https://github.com/zarr-developers/VirtualiZarr/issues/559)
+- [Redefine Virtual Readers as func(filepath) -> ManifestStore](https://github.com/zarr-developers/VirtualiZarr/issues/498)
+- [Make xarray an optional dependency?](https://github.com/zarr-developers/VirtualiZarr/issues/521)
+- [Re-implement loadable_variables using ManifestStore ](https://github.com/zarr-developers/VirtualiZarr/issues/473)
+- [Should we try to use Obstore for reading files?](https://github.com/zarr-developers/VirtualiZarr/issues/476)
+- [Challenges with earthaccess.open_virtual_mfdataset for grouped netCDFs with nested structure](https://github.com/zarr-developers/VirtualiZarr/issues/487)
+- [Docs on how to write a custom reader](https://github.com/zarr-developers/VirtualiZarr/issues/452)
+- [Make readers pluggable via entrypoint system](https://github.com/zarr-developers/VirtualiZarr/issues/245)
+- [Make Manifests' in-memory reference structure pluggable](https://github.com/zarr-developers/VirtualiZarr/issues/246)
+- [Architecture diagram](https://github.com/zarr-developers/VirtualiZarr/issues/225)
+
+## Data structures
+
+- `ChunkManifest`: Contains paths, chunk byte offsets, lengths
+- `ManifestArray`: Contains one ChunkManifest and Zarr V3 ArrayMetadata
+- `ManifestGroup`: Contains group metadata and ManifestGroup and/or ManifestArrays
+- `ManifestStore`: Contains a ManifestGroup and a FileReader
+- `xarray.Dataset`: Contains ManifestArray-backed variables
+
+## Protocols
+
+> **Note:** Protocols over ABCs would allow third party backends to be less tightly coupled to VirtualiZarr
+>
+> **Question:** How do we simplify these protocol definitions using obspec?
+
+- `FileReader`: read bytes from files on object storage or disk
+- `VirtualZarrBackend`: construct a ManifestStore from a file
+
+```python
+class FileReader(Protocol):
```

Comment:
This seems like a useful protocol, but it shouldn't live in VirtualiZarr, as it's a much more generally-useful thing. Xarray could use this, for example.